Fix/python ensurepip wheels

smoser opened the pull request:
This does:
1. update python-3.12 to 3.12.5 which released yesterday
2. Add a package `ensurepip-wheels` which includes wheels for setuptools and pip and installs them in /usr/share/python-wheels
3. compile python with `--with-wheel-pkg-dir=/usr/share/python-wheels` which how fedora and ubuntu build, and is described [here](https://github.com/python/cpython/blob/3e753c689a802d2e6d909cce3e22173977b2edbf/Lib/ensurepip/__init__.py#L15)
4. Replace the included wheels (ensurepip/_bundled) in python-3.10 and 3.11 with a dependency on ensurepip-wheels

Things to note:
* python3.12 no longer has setuptools included [95299](https://github.com/python/cpython/issues/95299).
* python3.12 is at least currently updating the bundled pip wheel.

smoser committed:
python 3.11 - Remove bundled python deps, replace with ensurepip-wheels dep

Rather than using the bundled wheels, we can bump our own.
The with-wheel-pkg-dir option was kind of made for this.

smoser committed:
python-3.10 - Remove bundled python deps, replace with ensurepip-wheels

Rather than using the bundled wheels, we can bump our own.
The with-wheel-pkg-dir option was kind of made for this.

xnox commented on python-3.12.yaml:
```diff
@@ -48,7 +48,7 @@ var-transforms:
     with:
-      expected-commit: 8e8a4baf652f6e1cee7acde9d78c4b6154539748
+      expected-commit: ff3bc82f7c9882c27aad597aac79355da7257186
       repository: https://github.com/python/cpython.git
```
@smoser Why no --with-wheel-pkg-dir=/usr/share/python-wheels here in python-3.12?

smoser replied:
> Things to note:
>
> *  python3.12 no longer has setuptools included https://github.com/python/cpython/issues/95299.
> *  python3.12 is at least currently updating the bundled pip wheel.

I'm happy to do that if you think we should. and just not add the ensurepip-wheels dep?

xnox replied:
my head hurts, one of those ways is deprecated right?!

pnasrat replied:
No. It's not related to depreciation as setup tools is still a build backend. For now not doing 3.12 is what I suggested I wouldn't block getting the older versions fixed here

pnasrat reviewed (dismissed):
LGTM

pnasrat commented:
@smoser test failures ?

smoser commented:
> my head hurts, one of those ways is deprecated right?!

Below is my plan. I want to call out the potential fallout here first.
**This plan could break things**. 
A package that successfully built in a venv previously because the newly created venv will now have a newer version of setuptools.


There is no indication that `--with-wheel-pkg-dir` is deprecated.  It is present in current upstream [Lib/ensurepip/__init__.py](https://github.com/python/cpython/blob/540fcc62f5da982b79504221cac01bfab8b73ba1/Lib/ensurepip/__init__.py#L12). 

Based on https://github.com/python/cpython/issues/122744 issue, it seems that upstream is upgrading the bundled pip in 3.12 and main, but is not doing so for other supported python versions (3.8, 3.9, 3.10, 3.11).

I think the whole fix is to:
1. update to python 3.12.5, which contains pip at current latest version 24.2
2. Update all versions of python to use `--with-wheel-pkg-dir=/usr/share/python-wheels`
3. create 2 packages (py3-distwheel-setuptools and py3-distwheel-pip) that have proper `update` sections
4. Update 3.8...3.11 with dependency on py3-distwheel-setuptools and py3-distwheel-pip
5. update 3.12 with dependency on py3-distwheel-pip
6. rm -R ensurepip/_bundled in all python versions.


For some information, here is _current_ version info:


| pyver       | pip    | setuptools |
| ----------- | ----   | ---------- |
| 3.8.19-r1   | 23.0.1 | 56.0.0  |
| 3.9.19-r2   | 23.0.1 | 58.1.0  |
| 3.10.14-r4  | 23.0.1 | 65.5.0  |
| 3.11.9-r6   | 24.0   | 65.5.0  |
| 3.12.4-r1   | 24.0   |  [gone](https://github.com/python/cpython/issues/95299)   |


Below is how this was captured.

```
$ vers="3.8 3.9 3.10 3.11 3.12"
$ for v in $vers; do apk add python-$v-base; done >out 2>&1
$ grep Installing.*python- out
(13/13) Installing python-3.8-base (3.8.19-r1)
(1/1) Installing python-3.9-base (3.9.19-r2)
(1/1) Installing python-3.10-base (3.10.14-r4)
(1/1) Installing python-3.11-base (3.11.9-r6)
(1/1) Installing python-3.12-base (3.12.4-r1)

$ (set -x; for v in $vers; do python$v -m venv $d/$v; done)
+ python3.8 -m venv /tmp/tmp.XK9tsQ/3.8
+ python3.9 -m venv /tmp/tmp.XK9tsQ/3.9
+ python3.10 -m venv /tmp/tmp.XK9tsQ/3.10
+ python3.11 -m venv /tmp/tmp.XK9tsQ/3.11
+ python3.12 -m venv /tmp/tmp.XK9tsQ/3.12

$ (set -x; for v in $vers; do $d/$v/bin/pip list; done)
+ /tmp/tmp.XK9tsQ/3.8/bin/pip list
Package    Version
---------- -------
pip        23.0.1
setuptools 56.0.0

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: python3.8 -m pip install --upgrade pip
-----

+ /tmp/tmp.XK9tsQ/3.9/bin/pip list
Package    Version
---------- -------
pip        23.0.1
setuptools 58.1.0

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: python3.9 -m pip install --upgrade pip
-----

+ /tmp/tmp.XK9tsQ/3.10/bin/pip list
Package    Version
---------- -------
pip        23.0.1
setuptools 65.5.0

[notice] A new release of pip is available: 23.0.1 -> 24.2
[notice] To update, run: python3.10 -m pip install --upgrade pip
-----

+ /tmp/tmp.XK9tsQ/3.11/bin/pip list
Package    Version
---------- -------
pip        24.0
setuptools 65.5.0

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python3.11 -m pip install --upgrade pip
-----

+ /tmp/tmp.XK9tsQ/3.12/bin/pip list
Package Version
------- -------
pip     24.0

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python3.12 -m pip install --upgrade pip


```

smoser commented:
> @smoser test failures ?

I have the fix for that, 3.10 was missing a dep.

smoser commented:
I would like to consider this ready, of the full plan, it would leave:

> 1. :heavy_check_mark: update to python 3.12.5, which contains pip at current latest version 24.2
> 2. :heavy_check_mark: Update all versions of python to use --with-wheel-pkg-dir=/usr/share/python-wheels

those are done.

> 3. [ ] create 2 packages (py3-distwheel-setuptools and py3-distwheel-pip) that have proper update sections
> 4. [ ] Update 3.8...3.11 with dependency on py3-distwheel-setuptools and py3-distwheel-pip

Currently this just has the 'ensurepip-wheels' package as a dep.

> 5. [ ] update 3.12 with dependency on py3-distwheel-pip

3.12 does not have any depenedencies but has up to date pip and no setuptools.

> 6. [ ] rm -R ensurepip/_bundled in all python versions.

This is done in all but 3.12

pnasrat reviewed (dismissed):
LGTM

smoser committed:
Add py3-pip-wheel and py3-setuptools-wheel as for ensurepip

cpython's stdlib package 'ensurepip' bundles a pip wheel inside
the python distribution.  That is used to 'ensure' that there
is a pip inside a `python3 -m venv mydir`.  In python 3.8 to
python 3.11 there is also a bundled setuptools wheel.

That wheel is updated occasionally.

In order to make sure we're shipping the most recent versions
of setuptools and pip, we are removing the wheels from the
python packages and adding a dependency on these "wheel"
packages.

smoser committed:
python-3.12/3.12.5 package update

smoser committed:
python-3.12 - Remove bundled python deps, replace with ensurepip-wheels

Rather than using the bundled wheels, we can bump our own.
The with-wheel-pkg-dir option was kind of made for this.

python-3.12+ does not include bundled setuptools wheel, only pip.

smoser commented:
All of the 6 items listed above are done in this pr right now.

smoser commented:
It appears that this is causing ordering issues for wolfictl.
To avoid those issues I put up https://github.com/wolfi-dev/os/pull/25860 .
We can land that, and then this will  build correctly (or i can drop the commits that add the wheel packages)

pnasrat reviewed (approved):
LGTM

xnox committed:
Merge branch 'main' into fix/python-ensurepip-wheels
